joblib/numpy_pickle.py

```diff
@@ -1,224 +1,231 @@
+        """Read the array corresponding to this wrapper.
 
-    #def __reduce__(self):
-    #    return None
+        Use the unpickler to get all information to correctly read the array.
 
+        Parameters
+        ----------
+        unpickler: NumpyUnpickler
 
-class ZNDArrayWrapper(NDArrayWrapper):
-    """An object to be persisted instead of numpy arrays.
-
-    This object store the Zfile filename in which
-    the data array has been persisted, and the meta information to
-    retrieve it.
-
-    The reason that we store the raw buffer data of the array and
-    the meta information, rather than array representation routine
-    (tostring) is that it enables us to use completely the strided
-    model to avoid memory copies (a and a.T store as fast). In
-    addition saving the heavy information separately can avoid
-    creating large temporary buffers when unpickling data with
-    large arrays.
-    """
-    def __init__(self, filename, init_args, state):
-        "Store the useful information for later"
-        self.filename = filename
-        self.state = state
-        self.init_args = init_args
+        Returns
+        -------
+        array: numpy.ndarray
 
-    def read(self, unpickler):
-        "Reconstruct the array from the meta-information and the z-file"
-        # Here we a simply reproducing the unpickling mechanism for numpy
-        # arrays
-        filename = os.path.join(unpickler._dirname, self.filename)
-        array = unpickler.np.core.multiarray._reconstruct(*self.init_args)
-        with open(filename, 'rb') as f:
-            data = read_zfile(f)
-        state = self.state + (data,)
-        array.__setstate__(state)
-        return array
+        """
+        # When requested, only use memmap mode if allowed.
+        if unpickler.mmap_mode is not None and self.allow_mmap:
+            array = self.read_mmap(unpickler)
+        else:
+            array = self.read_array(unpickler)
 
+        # Manage array subclass case
+        if (hasattr(array, '__array_prepare__') and
+            self.subclass not in (unpickler.np.ndarray,
+                                  unpickler.np.memmap)):
+            # We need to reconstruct another subclass
+            new_array = unpickler.np.core.multiarray._reconstruct(self.subclass,
+                                                                  (0,), 'b')
+            return new_array.__array_prepare__(array)
+        else:
+            return array
 
 ###############################################################################
 # Pickler classes
 
+
 class NumpyPickler(Pickler):
     """A pickler to persist of big data efficiently.
 
-        The main features of this object are:
+    The main features of this object are:
+    * persistence of numpy arrays in a single file.
 
-         * persistence of numpy arrays in separate .npy files, for which
-           I/O is fast.
+    * optional compression using BinaryZlibFile, with a special care on avoiding
```

Comment:
Is the compression always using BinaryZlibFile. Maybe just say "optional compression with ..."